docs/design_proposals/event-api-v2.md

```diff
@@ -0,0 +1,119 @@
+
+To address this, **I propose that we change the Build() interface to return a list of build results, one for each artifact, with information about the original artifact skaffold attempted to build, and either the built artifact, or the error the builder encountered.**
+
+This will allow the runner to:
+* know which artifacts were built successfully, and which failed
+* log events for each artifact, complete with a detailed error message passed back from the runner in the event that the build didn't complete successfully
+
+With this information, we could move all of the event handling code into a decorator that we wrap the builders and deployers in during the creation of the runner, similar to the way we handle timings currently. The semantics for each state (Not Started, In Progress, Complete, Failed) and the events that transition the states should all be the same between builders and deployers, and we can share the logic for handling each of these events in a separate location. This has the added bonus of not requiring any event-related code to be written by contributors if/when they add new builders/deployers, or make changes to existing ones.
+
+
+## Implementation plan
+
+Implementing this should be pretty straightforward:
+
+1) **Rework the Build() interface**
+
+The current method signature is 
+
+```golang
+Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+```
+
+This will be changed to return a list of build results for each artifact, along with the artifact information itself. The artifact information can be embedded in the build result:
+
+```golang
+type BuildResult struct {
+  Target *latest.Artifact
+  Result *build.Artifact
+  Error error
+}
+```
+and the new method signature will be
+```golang
+Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]BuildResult, error)
+```
+
+2) **Consume the new build results in the runner**
+
+The runner can then take control of all the eventing code. Deploy results are already returned by the Deploy() interface implementations, so with the Build results we have all the information we need. Using Build() as an example (Deploy() will look similar):
+
+* For each artifact provided to Build(), an `BuildInProgress` event is triggered
+* Once the build is complete, for each returned result, trigger a `BuildFailed` event for results with errors, and a `BuildComplete` event for those without.
+
+```golang
+for _, a := range artifactsToBuild {
+  event.BuildInProgress(a)
+}
+bRes, err := r.Build(ctx, out, tags, artifactsToBuild)
+for _, res := range bRes {
+  if res.Error != nil {
+    event.BuildFailed(res.Target, res.Error)
+  } else {
+    event.BuildComplete(res.Target)
+  }
+}
+```
+
+The main drawback here is that for builds run in parallel, the statuses will not truly reflect reality until all builds are completed, but I don't believe users will see much of an effect from this.
```

Comment:
I agree.
- Most of the time, the `artifactsToBuild` will just have one item
- And even if there are more items, builds will usually succeed, which means the users are mainly interested in "when is the whole build cycle complete".

Reply:
umm i kind of disagree. 
For microservices, usually # of artifactsToBuild is > 1.

Can we instead use `go channels`? 
`Build`, `Deploy` will now consume a channel to write events to.
The runner will emit events that are written to the channel?

Reply:
@tejal29 I may be mistaken, but usually the build is triggered by the watcher, right? That means that the reason was some user change. And users usually make small confined changes in one source code file. Therefore I think that the queue should most of the time have just one artifact (the one that edited file belongs to).
This of course does not apply to the first run where all artifacts are built.

Reply:
@corneliusweig good point. For dev loop this would not affect much.
Builds can be triggered by
1. skaffold build
2. skaffold run
3. skaffold dev
In first 2 cases, parallel builds happen and events will be triggered at the end of all builds.
For dev, 
Ideally,  one file should be part of one artifact, however consider an bazel example.
```
apiVersion: skaffold/v1beta8
kind: Config
build:
  artifacts:
  - image: gcr.io/k8s-skaffold/skaffold-bazel-frontend
    bazel:
      target: //:skaffold_example-frontend.tar
  - image: gcr.io/k8s-skaffold/skaffold-bazel-backend
    bazel:
      target: //:skaffold_example-backend.tar
```
The targets `skaffold_example-frontend` and `skaffold_example-backend` depend on target `//common-lib`
In this case if user changes a file in common-lib, both of the images should trigger a build

Reply:
Yeah, this is the key contentious part in this redesign. 
`go channels` would be definitely a correct solution to stream build events - but it feels like a very large scale change. That could be a follow-up - or maybe we can totally punt it until we see evidence around needing it - and go with the @corneliusweig's assumption that I think can be true as well that - except the first build - most of the actual builds are going to be single artifact builds. 
Also, as I mentioned above, I agree that we should consult with the IDE teams on "batch" vs "streamed" build events.

Reply:
> go channels would be definitely a correct solution to stream build events - but it feels like a very large scale change.

I don't agree "a very large scale change" should be a measure to punt or shelf functionality to implement for later. 
We don't have evidence for our current assumption either which is "most of the times its a single artifact build"
Maybe our assumption is wrong. 

That being said, we need arrive to a reasonable decision especially in this scenario where do not have data on how users are using skaffold, how long their build are, and hence stakeholders come in to picture.
We do need to make it sure, the consumers of Event API know that these events are now going to be batched.

Reply:
For the record - I did change my mind on this and pushed for keeping the "realtime" behavior in our conversations.